src/GafferOSL/OSLShader.cpp

```diff
@@ -199,19 +204,20 @@ bool OSLShader::acceptsInput( const Plug *plug, const Plug *inputPlug ) const
 
 		if( sourceShaderOutPlug && ( sourceShaderOutPlug == inputPlug || sourceShaderOutPlug->isAncestorOf( inputPlug ) ) )
 		{
-			// Source is the output of a shader node, so it needs to
-			// be coming from another OSL shader. Although because Arnold allows
-			// mixing and matching Arnold native shaders and OSL shaders,
-			// we must also accept connections from ArnoldShaders.
-			if( !sourceShader->isInstanceOf( staticTypeId() ) && !sourceShader->isInstanceOf( "GafferArnold::ArnoldShader" ) )
+			if( sourceShader->isInstanceOf( staticTypeId() ) )
 			{
-				return false;
+				return true;
 			}
-			const std::string sourceShaderType = sourceShader->typePlug()->getValue();
-			if( sourceShaderType != "osl:shader" && sourceShaderType != "ai:surface" )
```

Comment:
When we spoke about this before, I hadn't realised we also had this check that we've now lost. OSL is moving away from giving any importance to the shader type, so it might be that we no longer care if an "osl:surface" is plugged in instead of an "osl:shader". But since ArnoldShader is used to load "ai:light" and "ai:color_manager" we should probably continue to reject those. Perhaps we should stop caring about `typeId()` and have `registerCompatible()` shader take strings like "ai:surface" etc? Would that work for your use case?

Reply:
Sure I don't mind either way, I'll set that up instead.

Reply:
Just to note, perhaps wildcards might work here eg. "ccl:*" as I think all the types in Cycles can have OSL attached, but it's no big deal, I can just register a handful of them if needed.